### PR TITLE
Feature/prepare xively demo application for stm32f4 nucleo

### PR DESCRIPTION
**Pull request by olgierdh**
[DESCRIPTION]

Template for the demo application for STMFx based devices that uses Xively. 

[REVIEWERS]

[QA]

Tested on STMF4xxx with X-NUCLEO-IKS01A.

### CODE REVIEW

**Comment by Palantir555** [comment]
The code looks great! I think we're good to merge :ok_hand:

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 1]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
+
+/* table of topics used for this demo */
+static const mqtt_topic_configuration_t topics_array[] = {
+	{{ XI_TOPIC_NAME_MANGLE( "Accelerometer" ), MQTT_TOPIC_PUB }, NULL, pub_accelerometer },
+	{{ XI_TOPIC_NAME_MANGLE( "Barometer" ), MQTT_TOPIC_PUB }, NULL, pub_barometer },
+	{{ XI_TOPIC_NAME_MANGLE( "Button" ), MQTT_TOPIC_PUB }, NULL, pub_button },
+	{{ XI_TOPIC_NAME_MANGLE( "Gyroscope" ), MQTT_TOPIC_PUB }, NULL, pub_gyroscope },
+	{{ XI_TOPIC_NAME_MANGLE( "Humidity" ), MQTT_TOPIC_PUB }, NULL, pub_humidity },
+	{{ XI_TOPIC_NAME_MANGLE( "LED" ), MQTT_TOPIC_SUB }, &on_led_msg, NULL },
+	{{ XI_TOPIC_NAME_MANGLE( "Temperature" ), MQTT_TOPIC_PUB }, NULL, pub_temperature },
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: +Magnetometer? Yeah, I know it is not yet in the PL.

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 2]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
+void push_mqtt_topics( )
+{
+	int i = 0;
+	for(; i < topics_array_length; ++i )
+	{
+		const mqtt_topic_configuration_t* const topic_config = &topics_array[ i ];
+
+		if( MQTT_TOPIC_PUB == topic_config->topic_descr.topic_type )
+		{
+			if( NULL != topic_config->on_msg_push )
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: can't the existence of the on_msg_push define that this topic_config is a PUB type? I mean isn't this representation redundant?

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 3]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
+	return xi_publish( gXivelyContextHandle,
+			mqtt_topic_descr->name, "greetings from barometer",
+			XI_MQTT_QOS_AT_MOST_ONCE,
+			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+}
+
+xi_state_t pub_button( const mqtt_topic_descr_t* const mqtt_topic_descr )
+{
+	return xi_publish( gXivelyContextHandle,
+			mqtt_topic_descr->name, "greetings from button",
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: greetings dear button

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 4]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
     printf( "\r\nconnected, state : %i" , conn_data->connection_state  );
+	switch ( conn_data->connection_state )
+	{
+		/* connection attempt failed */
+		case XI_CONNECTION_STATE_OPEN_FAILED:
+			printf( "[%d] Xively: Connection failed %s:%d, error %d, reconnecting....\r\n",
+					( int ) xi_bsp_time_getcurrenttime_seconds(),
+					conn_data->host, conn_data->port, state );
+
+			xi_connect( in_context_handle,
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: you will connect you bad board!!

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 5]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
+		sizeof( topics_array ) / sizeof( topics_array[ 0 ] );
+
+xi_context_handle_t gXivelyContextHandle = XI_INVALID_CONTEXT_HANDLE;
+xi_timed_task_handle_t gXivelyTimedTaskHandle = XI_INVALID_TIMED_TASK_HANDLE;
+
+
+/* handler for led msg */
+void on_led_msg( const uint8_t* const msg, size_t msg_size )
+{
+	printf( "on_led_msg handler for the LED topic\r\n" );
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: :boom:

File: src/bsp/platform/stm32fx_nucleo_wifi/xi_bsp_io_net_stm32fx_nucleo_wifi.c [hunk 6]
```diff
@@ -242,7 +242,7 @@ void xi_bsp_io_net_socket_data_received_proxy( uint8_t socket_id,
         return;
     }
 
-    XI_LIST_PUSH_FRONT( xi_data_desc_t, xi_net_state.head, tail );
+    XI_LIST_PUSH_BACK( xi_data_desc_t, xi_net_state.head, tail );
```
Comment (atigyi) [inline src/bsp/platform/stm32fx_nucleo_wifi/xi_bsp_io_net_stm32fx_nucleo_wifi.c]: oh wow :+1:

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 7]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
 char * seckey = "PASSWORD";
 WiFi_Priv_Mode mode = WPA_Personal;
 
-char* XI_ACCOUNT_ID = "XIVELY ACCOUNT ID";
-char* XI_USER = "XIVELY USER";
-char* XI_DEVICE_ID = "DEVICE ID";
-char* XI_DEVICE_PASS = "DEVICE PASSWORD";
+#define XI_ACCOUNT_ID "XIVELY ACCOUNT ID"
+#define XI_DEVICE_ID  "XIVELY DEVICE ID"
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: simpllification -XI_USER :+1:

File: examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c [hunk 8]
```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
...
+			if( XI_STATE_OK != ret )
+			{
+				return ret;
+			}
+		}
+	}
+
+	/* registration of the publish function */
+	gXivelyTimedTaskHandle = xi_schedule_timed_task( in_context_handle,
+			&push_mqtt_topics, XI_PUBLISH_INTERVAL_SEC, 1, NULL );
```
Comment (atigyi) [inline examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c]: that is why we do not need RTOS and threads

**Comment by atigyi** [comment]
Awesomeness, I will miss these professional moves :cry: 
One request for simplification for consideration, otherwise :100: / :100:

**Comment by olgierdh** [comment]
Thanks! And I will miss yours 😭  I was thinking about making use of these pointers instead of the PUB SUB types - I will fix that - remove the redundancy - thanks!

File: src/bsp/platform/stm32fx_nucleo_wifi/xi_bsp_io_net_stm32fx_nucleo_wifi.c (hunk 6 above)
Comment (olgierdh) [inline src/bsp/platform/stm32fx_nucleo_wifi/xi_bsp_io_net_stm32fx_nucleo_wifi.c]: Yeah... heh.

**Pull request merged**